### ISSUE TITLE
`torch.compile` (inductor) — `*_like` ops don't preserve a size-1 dimension's stride under `preserve_format` (eager does)

### ISSUE THREAD

**Issue by tinywisdom** [pytorch/pytorch#197815]
### 🐛 Describe the bug

The `*_like` factory ops (`ones_like`, `zeros_like`, `empty_like`, `rand_like`, `full_like`)
default to `memory_format=torch.preserve_format`, i.e. the output should keep the input tensor's
strides. Under `torch.compile` (inductor), when the input has a **size-1 dimension** with a
non-standard stride, the compiled output does not preserve that dimension's stride (eager does),
so eager and compiled disagree on the output's strides. The values are correct; only the stride
metadata differs.

## Minimal repro

```python
import torch

x = torch.randn(1, 5, device='cuda').t()   # shape (5, 1), stride (1, 5); 5 elements

def fn(t):
    return torch.ones_like(t)               # default memory_format=preserve_format

print(x.stride())                                        # (1, 5)
print(fn(x).stride())                                    # eager:    (1, 5)  -- size-1 dim stride preserved
print(torch.compile(fn, backend="inductor")(x).stride()) # compiled: (1, 1)  -- size-1 dim stride changed
```

The whole `*_like` family is affected, always on a **size-1** dimension:

| input | numel | eager stride | compiled stride |
|---|---|---|---|
| `randn(1,5).t()` → (5,1) | 5 | (1, 5) | (1, **1**) |
| `randn(2,4,1).transpose(1,2)` → (2,1,4) | 8 | (4, 1, 1) | (4, **4**, 1) |
| `randn(4,1,5).permute(1,0,2)` → (1,4,5) | 20 | (5, 5, 1) | (**20**, 5, 1) |
| `randn(3,1,4).transpose(0,1)` → (1,3,4) | 12 | (4, 4, 1) | (**12**, 4, 1) |

Genuine permutations with **no size-1 dim** (`randn(3,4).t()`, a channels-last 4-D tensor,
`permute(2,0,1)`) are preserved correctly by both, so the gap is specific to the size-1 dimension.

## Expected behavior

With `memory_format=preserve_format` (the default), the compiled output should keep the input's
strides — including the stride of a size-1 dimension — matching eager.

## Actual behavior

Eager preserves the size-1 dimension's stride (e.g. `(1, 5)`); inductor produces a normalized
stride (e.g. `(1, 1)`). Values are identical; the output strides differ. The differing stride is
always on a size-1 dimension, so element addressing is unaffected — this is a metadata-only
divergence, but it is still a `preserve_format` mismatch between eager and compiled.

### Versions

```
PyTorch version: 2.15.0a0+gite67e43b
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04.3) 12.3.0
Clang version: 17.0.6
CMake version: version 4.0.0
Libc version: glibc-2.35

Python version: 3.12.13 | packaged by conda-forge (64-bit runtime)
Python platform: Linux-6.8.0-138-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
GPU 0: NVIDIA RTX A6000
Nvidia driver version: 580.82.07
cuDNN version: 9.3.0

[pip3] numpy==2.3.5
[pip3] torch==2.15.0a0+gite67e43b
[pip3] triton==3.7.1
```

cc @jamesr66a @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by indrajeetapache** [comment]
Working on a fix for this in https://github.com/pytorch/pytorch/pull/197831

**Why it breaks:** with `preserve_format` (the `*_like` default), eager ATen does `empty_strided(sizes, strides)` for non-overlapping dense inputs, so a size-1 dim keeps its real stride (e.g. `(5,1)` / `(1,5)`). Inductor/refs instead rebuild layout through `compute_elementwise_output_logical_to_physical_perm`. Contiguity **skips size-1 dims**, so that tensor looks contiguous and the output gets canonical strides (`(1,1)`). Values stay correct; only stride metadata diverges.

**What the fix does:**
1. In refs `empty_like`, if the input is non-overlapping and dense, allocate with `empty_strided(shape, stride)` (same as ATen).
2. In Inductor `full_like` / `_rand_like`, same dense short-circuit: fill/random into a contiguous buffer, then `copy` into an `empty_strided` with the input strides when they differ.

That covers `ones_like` / `zeros_like` / `empty_like` / `full_like` / `rand*_like` under `preserve_format` for the size-1 stride cases in the issue.
